### PR TITLE
FR-1524: Trigger Additional Hearing Document

### PR DESCRIPTION
**Pull request by HarryH96**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FR-1524

### Change description ###
- Trigger Hearing Document when Form C and G exists
- Send this document to applicant and respondent, with cover sheet, for bulk printing

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/controllers/HearingDocumentController.java [hunk 1]
```diff
@@ -45,7 +49,7 @@
             @ApiResponse(code = 500, message = "Internal Server Error")})
     public ResponseEntity<AboutToStartOrSubmitCallbackResponse> generateHearingDocument(
             @RequestHeader(value = AUTHORIZATION_HEADER) String authorisationToken,
-            @NotNull @RequestBody @ApiParam("CaseData") CallbackRequest callback) {
+            @NotNull @RequestBody @ApiParam("CaseData") CallbackRequest callback) throws IOException {
```
Comment (shaunganley) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/controllers/HearingDocumentController.java:52]: not sure we want to just throw exceptions here. at least catch the exception, log the error before returning an appropriate exception

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/controllers/HearingDocumentController.java [hunk 2]
```diff
@@ -63,8 +67,13 @@
         caseData.putAll(service.generateHearingDocuments(authorisationToken, caseDetails));
 
         if (isContestedPaperApplication(caseDetails)) {
-            log.info("Sending Contested Paper Case to bulk print Case ID: {}", caseDetails.getId());
-            service.sendToBulkPrint(caseDetails, authorisationToken);
+            if (isDocumentPresentInCaseData(FORM_C, caseDetails) && isDocumentPresentInCaseData(FORM_G, caseDetails)) {
```
Comment (shaunganley) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/controllers/HearingDocumentController.java:70]: move this logic to a well named method, something like hasHearingDocuments

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/CommonFunction.java [hunk 3]
```diff
@@ -136,4 +136,8 @@ private static boolean isConsentedApprovedOrderCollectionPresent(Map<String, Obj
     private static boolean isContestedApprovedOrderCollectionPresent(Map<String, Object> caseData) {
         return caseData.get(CONTESTED_CONSENT_ORDER_COLLECTION) != null && !((List<Map>) caseData.get(CONTESTED_CONSENT_ORDER_COLLECTION)).isEmpty();
     }
+
+    public static boolean isDocumentPresentInCaseData(String documentName, CaseDetails caseDetails) {
```
Comment (shaunganley) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/CommonFunction.java:140]: not sure this method really offers us anything. it shouldn't be specific to documents as the logic would work for any case data field. I think just call containsKey method directly

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java [hunk 4]
```diff
@@ -132,4 +147,82 @@ public void sendToBulkPrint(CaseDetails caseDetails, String authorisationToken)
...
+        FrcCourtDetails selectedFRCDetails = FrcCourtDetails.builder()
+            .courtName((String) courtDetails.get(COURT_DETAILS_NAME_KEY))
+            .courtAddress((String) courtDetails.get(COURT_DETAILS_ADDRESS_KEY))
+            .phoneNumber((String) courtDetails.get(COURT_DETAILS_PHONE_KEY))
+            .email((String) courtDetails.get(COURT_DETAILS_EMAIL_KEY))
+            .build();
+
+        caseData.put("ccdCaseNumber", caseDetails.getId());
+        caseData.put("DivorceCaseNumber", caseDetails.getData().get(DIVORCE_CASE_NUMBER));
+        caseData.put("ApplicantName", buildFullApplicantName(caseDetails));
```
Comment (shaunganley) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java:181]: we would get a lot of these values if we called the prepareLetterTemplateData method in DocumentHelper. I'm keen that all of our letters call this reusable method to add generic values like names, address details, court details and case numbers. then only add values that are specific to this document in here.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java [hunk 5]
```diff
@@ -132,4 +147,82 @@ public void sendToBulkPrint(CaseDetails caseDetails, String authorisationToken)
...
+    private void bulkPrintAdditionalHearingDocuments(CaseDetails caseDetails, String authorisationToken) {
+        List<AdditionalHearingDocumentData> additionalHearingDocumentData =
+            documentHelper.convertToAdditionalHearingDocumentData(
+                caseDetails.getData().get(ADDITIONAL_HEARING_DOCUMENT_COLLECTION));
+
+        AdditionalHearingDocumentData additionalHearingDocument = additionalHearingDocumentData.get(additionalHearingDocumentData.size() - 1);
+
+        List<BulkPrintDocument> document = Collections.singletonList(
+            bulkPrintService.getBulkPrintDocumentFromCaseDocument(
+                additionalHearingDocument.getAdditionalHearingDocument().getAdditionalHearingDocument()));
```
Comment (shaunganley) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java:223]: this looks like we're using complex types where we shouldn't be. we should really only have to do AdditionalHearingDocumentData. getAdditionalHearingDocument() here

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java (hunk 4 above)
Comment (HarryH96) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java:181]: Initially I had used this, however as the dcoument is actually not a letter I decided against using these methods and having them in my own prep method

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java (hunk 5 above)
Comment (HarryH96) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/HearingDocumentService.java:223]: I think this is a poor naming convention on my part. It is now:
additionalHearingDocument.getAdditionalHearingDocument().getDocument()

If you check AdditionalHearingDocumentData.java, the variable is called additionalHearingDocument and in the AdditionalHearingDocument.java it is the same name! I will updated this in my next push

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java [hunk 6]
```diff
@@ -54,7 +54,7 @@ public void setUp() {
         config.setAdditionalHearingFileName("AdditionalHearingDocument.pdf");
     }
 
-    @Test
+    @Test(expected = IllegalStateException.class)
```
Comment (johnoneill46) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java:57]: 💣

**Review by johnoneill46 — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java (hunk 6 above)
Comment (HarryH96) [reply src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java:57]: Just trying to get a green build for now, will remove before merge!

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/CommonFunction.java (hunk 3 above)
Comment (shaunganley) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/CommonFunction.java:140]: still don't think this offers us anything.

File: src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        assertThat(data.get("CourtPhone"), is("0115 910 3504"));
+        assertThat(data.get("CourtEmail"), is("FRCNottingham@justice.gov.uk"));
+
+        verify(bulkPrintService, times(1)).printApplicantDocuments(eq(caseDetails), eq(AUTH_TOKEN), bulkPrintDocumentsCaptor.capture());
+        verify(bulkPrintService, times(1)).printRespondentDocuments(eq(caseDetails), eq(AUTH_TOKEN), bulkPrintDocumentsCaptor.capture());
+        assertThat(bulkPrintDocumentsCaptor.getValue().size(), is(1));
+    }
+
+
+    //TODO: Test that the document is successfully added to the case data
```
Comment (shaunganley) [inline src/test/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/AdditionalHearingDocumentServiceTest.java:97]: looks like you've more scenarios to cover here

**Review by shaunganley — CHANGES_REQUESTED** [review]
some small comments

**Review by shaunganley — APPROVED** [review]

**Pull request merged**